### PR TITLE
Replace several Ops by OpFromGraph

### LINKED ISSUE #1285: Add option not to print inner graphs in debug_print
### Description

...

### LINKED ISSUE #2106: BUG: UserWarning: Numba will use object mode to run XlogY0'
### Describe the issue:

When calling some preliz functions (which use pytensor-distributions), we got 

UserWarning: Numba will use object mode to run XlogY0's perform method. Set `pytensor.config.compiler_verbose = True` to see more details.
  warnings.warn(

### Reproducable code example:

```python
import preliz as pz
pz.Beta(2, 5).pdf(0.1)
```

### Error message:

```shell

```

### PyTensor version information:

....

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by ricardoV94**
- **Add print_inner_graphs parameter to debugprint** Closes #1285 
- **Add SymbolicOp: OpFromGraph subclass with deferred graph construction**
- **Convert AllocDiag and KroneckerProduct to SymbolicOp**
- **Implement xlogy and xlog1py as SymbolicOp, deprecate xlogx module**  Closes #2106 
- **Convert Softmax and LogSoftmax to TensorSymbolicOp**

Related to https://github.com/pymc-devs/pytensor/issues/1210

These Ops are needed to hide the inner graph for easier graph rewriting, or to control autodiff. There's no need to implement new primitives for this functionality. Removing primitive Ops makes maintaining multiple backends easier, and benefits from optimizations we do do the core primitives.

I had postponed softmax because it had a custom C kernel, but with the change from it being the default I think it's fine. Numba is doing the same thing it was doing before (but less code), and libraries that may prefer direct dispatch like JAX/PyTorch/MLX continue to do so.

 I adde a new SymbolicOp subclass, much in line with the prototype SymbolicRandomVariable in PyMC. It has a method to build the inner graph on demant from input types, and a simpler interface than OFG (OFG is a bit too much already, so I didn't want to extend it further, we can reconsider this).

Next we could tackle https://github.com/pymc-devs/pytensor/issues/1221, which would allow us to still work with them as objects (needed for dispatching in PyMC), but allows us to have less real Ops, and to also inline, which can be great for memory optimization / rewrites. (E.g., a MvNormal L -> covariance -> L across the RV boundary doesn't get optimized now), plus memory opt and all that. That's left for another PR though

### CODE REVIEW

File: tests/tensor/test_special.py [hunk 1]
```diff
@@ -244,3 +212,82 @@ def test_betaln():
...
+    assert gxv[0] == -np.inf
+    assert np.isnan(gyv[0])
+
+    # x=0, y=1: grad_x = log(1) = 0, grad_y = 0/1 = 0
+    gxv, gyv = f(np.array([0.0]), np.array([1.0]))
+    np.testing.assert_allclose(gxv, 0.0)
+    np.testing.assert_allclose(gyv, 0.0)
+
+
+def test_xlogy_as_xlogx():
```
Comment (jessegrabowski) [inline tests/tensor/test_special.py]: probably don't need this one

**Review by jessegrabowski — APPROVED** [review]
very cool, seems to simplify things a lot.

If I understand well, this is new SymbolicTensorOp should be used in cases when I just want to "mark" a block of computation for rewrites to reason about jointly? What is the use-case for standard OpFromGraph after this?

**Comment by ricardoV94** [comment]
> If I understand well, this is new SymbolicTensorOp should be used in cases when I just want to "mark" a block of computation for rewrites to reason about jointly? What is the use-case for standard OpFromGraph after this?

This is just syntactic sugar of top of OFG with nicer API imo. If you want it to work with different types (vector, matrix, etc) you give a generative function for it. The big difference is this function lives in the Op class, whereas before it used to be a distinct floating helper.

This is pretty much the thing that SymbolicRV evolved into, so it's the product of some learning of what's needed in practice.

As I said this could be merged into OFG (or one day replace OFG), but it already has so much complex API I didn't want to burden it even more.

**Comment by ricardoV94** [comment]
> If I understand well, this is new SymbolicTensorOp should be used in cases when I just want to "mark" a block of computation for rewrites to reason about jointly? 

That and custom gradients. It's the reason why xlogy and xlogyp1 are not just helpers that build the forward graph

File: tests/tensor/test_special.py (hunk 1 above)
Comment (ricardoV94) [reply tests/tensor/test_special.py]: It's good to confirm we're not being bitten by that repeated input in the inner graph gotcha.

Comment (ricardoV94) [reply tests/tensor/test_special.py]: but the tests can certainly be merged
